dm/dm/master/shardddl/optimist.go

```diff
@@ -134,31 +139,162 @@ func (o *Optimist) ShowLocks(task string, sources []string) []*pb.DDLLock {
+						continue
+					}
+					if info, ok := ifm[lock.Task][source][schema][table]; ok {
+						owners = append(owners, utils.GenDDLLockID(source, schema, table))
+						ddlGroups = append(ddlGroups, info.DDLs)
+					}
+				}
+			}
+		}
+		if opms, ok := opm[lock.Task]; ok {
+			if len(sources) > 0 {
+				for _, source := range sources {
+					if opmss, ok := opms[source]; ok {
+						appendOwnerDDLs(opmss, source)
+					}
+				}
+			} else {
+				for source, opmss := range opms {
+					appendOwnerDDLs(opmss, source)
+				}
+			}
 		}
+		lockSynced := make([]string, 0, len(ready))
+		lockUnsynced := make([]string, 0, len(ready))
 		for source, schemaTables := range ready {
 			for schema, tables := range schemaTables {
 				for table, synced := range tables {
 					if synced {
-						l.Synced = append(l.Synced, fmt.Sprintf("%s-%s", source, dbutil.TableName(schema, table)))
+						lockSynced = append(lockSynced, fmt.Sprintf("%s-%s", source, dbutil.TableName(schema, table)))
 					} else {
-						l.Unsynced = append(l.Unsynced, fmt.Sprintf("%s-%s", source, dbutil.TableName(schema, table)))
+						lockUnsynced = append(lockUnsynced, fmt.Sprintf("%s-%s", source, dbutil.TableName(schema, table)))
 					}
 				}
 			}
 		}
-		sort.Strings(l.Synced)
-		sort.Strings(l.Unsynced)
-		ret = append(ret, l)
+		sort.Strings(lockSynced)
+		sort.Strings(lockUnsynced)
+
+		if len(owners) == 0 {
+			owners = append(owners, "")
+			ddlGroups = append(ddlGroups, nil)
+		}
+		for i, owner := range owners {
+			ret = append(ret, &pb.DDLLock{
+				ID:       lock.ID,
+				Task:     lock.Task,
+				Mode:     config.ShardOptimistic,
+				Owner:    owner,
+				DDLs:     ddlGroups[i],
+				Synced:   lockSynced,
+				Unsynced: lockUnsynced,
+			})
+		}
 	}
-	return ret
+	return ret, err
+}
+
+// UnlockLock unlocks a shard DDL lock manually only when using `unlock-ddl-lock` command.
+// ID: the shard DDL lock ID.
+// source, upstreamSchema, upstreamTable: reveal the upstream table's info which we need to skip/exec
+// action: whether to skip/exec the blocking DDLs for the specified upstream table
+// NOTE: this function has side effects, if it failed, some status can't revert anymore.
+// NOTE: this function should not be called if the lock is still in automatic resolving.
+func (o *Optimist) UnlockLock(ctx context.Context, id, source, upstreamSchema, upstreamTable string, action pb.UnlockDDLLockOp) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if o.closed {
+		return terror.ErrMasterOptimistNotStarted.Generate()
+	}
+	task := utils.ExtractTaskFromLockID(id)
+	// 1. find the lock.
+	lock := o.lk.FindLock(id)
+	if lock == nil {
+		return terror.ErrMasterLockNotFound.Generate(id)
+	}
+
+	// 2. check whether has resolved before (this often should not happen).
+	if lock.IsResolved() {
+		deleted, err := o.removeLock(lock)
+		if err != nil {
+			return err
+		}
+		deletedStr := "deleted"
+		if !deleted {
+			deletedStr = "not " + deletedStr
```

Comment:
The status `not deleted` is a bit misleading because in most cast it is returned because the lock is not existed in etcd, but the error message looks like we are failed to delete the metadata in etcd.